Add TestPropertiesDrivenEncryption

andersonm-1 opened the pull request.

andersonm-ibm committed:
Add TestPropertiesDrivenEncryption

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+ * The read sample uses each of the following decryption configurations to read every
+ * encrypted files in the input directory:
+ *
+ *  - DECRYPT_COLUMNS_AND_FOOTER:   Decrypt using key retriever that holds the keys of
```
lets call it DECRYPT_WITH_KEY_RETRIEVER

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+ *
+ *  - DECRYPT_COLUMNS_AND_FOOTER:   Decrypt using key retriever that holds the keys of
+ *                                  two encrypted columns and the footer key.
+ *  - READ_PLAINTEXT:   Do not decrypt anything.
```
lets call it NO_DECRYPTION

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+  public ErrorCollector errorCollector = new ErrorCollector();
+
+  private static final Base64.Encoder encoder = Base64.getEncoder();
+  private static final byte[] FOOTER_ENCRYPTION_KEY = new String("0123456789012345").getBytes();
```
lets call these "master keys", FOOTER_MASTER_KEY, etc, to diff this from DEKs in previous tests

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+    LOG.error(fullErrorMessage);
+  }
+
+  public enum EncryptionConfiguration {
```
please move these enums to the beginning of the class, makes it easier to understand the rest

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+ *                                  - plaintext footer mode.
+ *  - ENCRYPT_COLUMNS_AND_FOOTER_CTR:   Encrypt two columns and the footer, with different
+ *                                  keys. Use the alternative (AES_GCM_CTR_V1) algorithm.
+ *  - WRITE_PLAINTEXT:   Do not encrypt anything
```
lets call it NO_ENCRYPTION

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+    SimpleGroupFactory f = new SimpleGroupFactory(schema);
+
+    for (Map.Entry<EncryptionConfiguration, Configuration> encryptionConfigurationEntry : encryptionPropertiesMap.entrySet()) {
+      KeyToolkit.removeCacheEntriesForToken(KeyToolkit.DEFAULT_ACCESS_TOKEN);
```
lets use removeAllCacheEntries? so no need to mention tokens.
please also add a comment here, and some other places that require explanataion

andersonm-1 replied:
the method is not public

ggershinsky replied:
added a public removeCacheEntriesForAllTokens

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+      if (null == conf) {
+        conf = new Configuration();
+      } else {
+
```
rm empty line

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+      for (int fileNum = 0; fileNum < listOfFiles.length; fileNum++) {
+        KeyToolkit.removeCacheEntriesForToken(KeyToolkit.DEFAULT_ACCESS_TOKEN);
+        Path file = new Path(listOfFiles[fileNum].getAbsolutePath());
+        if (!file.getName().endsWith("parquet.encrypted")) { // Skip non encrypted files
```
don't we want to test reading unencrypted files?

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+  final WriteSupport<Group> writeSupport = new GroupWriteSupport();
+
+  @Test
+  public void testWriteReadEncryptedParquetFiles() throws IOException {
```
not only encrypted

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+    LOG.info("======== testWriteReadEncryptedParquetFiles {} ========", rootPath.toString());
+    // This map will hold various encryption configurations.
+    Map<EncryptionConfiguration, Configuration> encryptionPropertiesMap = getHadoopConfigurationForEncryption();
+    testWriteEncryptedParquetFiles(rootPath, encryptionPropertiesMap);
```
not only encrypted

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+    testWriteEncryptedParquetFiles(rootPath, encryptionPropertiesMap);
+    // This map will hold various decryption configurations.
+    Map<DecryptionConfiguration, Configuration> decryptionPropertiesMap = getHadoopConfigurationForDecryption();
+    testReadEncryptedParquetFiles(rootPath, decryptionPropertiesMap);
```
not only encrypted

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/crypto/TestPropertiesDrivenEncryption.java:
```diff
@@ -0,0 +1,459 @@
+ *                                  keys. Use the alternative (AES_GCM_CTR_V1) algorithm.
+ *  - WRITE_PLAINTEXT:   Do not encrypt anything
+ *
+ *
```
we also need to add a check for writing key material in separate files: eg a loop on the test, with two iterations: a) default b) "encryption.key.material.internal.storage=false".

Plus two other loops: on "encryption.double.wrapping" [a) default b) false] and on ""encryption.wrap.locally" [a) default b) true]

ggershinsky replied:
and add a key rotation test..

andersonm-1 replied:
Changed the test to parameterized test. 
Key rotation test will be added later.

andersonm-ibm committed:
Rename some encryption/decryption configurations and make the test parameterized
to test combinations of isKeyMaterialExternalStorage, isDoubleWrapping, isWrapLocally.
Add RemoteKmsClient mock for remote wrapping.

andersonm-ibm committed:
Make common method setCommonKMSProperties and extract classname strings from classes

andersonm-ibm committed:
Change TestPropertiesDrivenEncryption to accomodate latest API changes.

andersonm-ibm committed:
Remove StringUtils
